msbuild/rest-api-client-generation/README.md

```diff
@@ -0,0 +1,272 @@
+
+Nowadays, an application which consumes RestApi is a very common scenario. We are going to generate the Rest API client automatically during build process. We will use [NSawg](https://docs.microsoft.com/aspnet/core/tutorials/getting-started-with-nswag?view=aspnetcore-6.0&tabs=visual-studio)
+
+Our app will be a console app, because it is simpler. The kind of app which use the Rest Api client is not important.
+The example will consume the public [Pet Store API](https://petstore.swagger.io), which publish the [OpenAPI spec](https://petstore.swagger.io/v2/swagger.json)
+
+If you are not clear on terms such as tasks, targets, properties, or runtimes, you could first check out the docs that explain these concepts, starting with the [MSBuild Concepts article](https://docs.microsoft.com/visualstudio/msbuild/msbuild-concepts).
+
+## Option 1: Use pre defined MSBuid Exec Task
+
+We will use the ["Exec" MSBuild task](https://docs.microsoft.com/dotnet/api/microsoft.build.tasks.exec?view=msbuild-17-netcore), which simply invokes the specified process with the specified arguments, waits for it to complete, and then returns True if the process completed successfully, and False if an error occurred.
+
+NSwag code generation is possible to be used from MSBuild, by [NSwag.MSBuild](https://github.com/RicoSuter/NSwag/wiki/NSwag.MSBuild)
+
+The complete code version is in this PetReaderExecTaskExample folder, you can download and take a look. Anyway, we are going to go through step by step and explain some concept on the way.
+
+- We are going to create a new console application on Visual Studio named PetReaderExecTaskExample. We use net6.0.
+- Create another project in the same solution: PetShopRestClient (This is going to contain the generated client as a Library). We use netstandard 2.1. The generated client doesn't compile on netstandard 2.0.
+- Go to the PetReaderExecTaskExample project, and add a project dependence to PetShopRestClient project.
+- On PetShopRestClient, include the following NuGet packages
+  - Nswag.MSBuild, it will allow us access to the code generator from MSBuild
+  - Newtonsoft.Json, it will be needed to compile the generated client
+  - System.ComponentModel.Annotations, it will be needed to compile the generated client
+- On PetShopRestClient, add a folder (named PetShopRestClient) for the code generation and delete the Class1.cs automatically generated.
+- Create a text file named petshop-openapi-spec.json (on root). We are going to add the OpenApi spec, please copy the content from [here](https://petstore.swagger.io/v2/swagger.json) inside the file. Why did we commit the spec instead of read it online?, we like repetitive build and depending only from the input, consuming directly the api could transform a build which works today to a build which fails tomorrow from the same source. The picture saved on petshop-openapi-spec.json will allow us to still have a version which builds even if the spec changes.
```

Comment:
```suggestion
- Create a text file named petshop-openapi-spec.json (on root). We are going to add the OpenApi spec, please copy the content from [here](https://petstore.swagger.io/v2/swagger.json) inside the file. Why did we commit the spec instead of reading it online?, we like a reproducible build that depends only from the input, consuming the API directly could transform a build which works today into a build which fails tomorrow from the same source. The picture saved on petshop-openapi-spec.json will allow us to still have a version which builds even if the spec changes.
```